### DISCUSSION TITLE
Auto Updates and Imports Fail

### DISCUSSION THREAD

**Discussion by AtrociKitty** [AndreWohnsland/CocktailBerry#183] [Need help] [answered]
After using the all in one script on a new Pi 4 B, version 1.35.1 was successfully installed. However, the auto update consistently fails, prompting to update at every boot (if enabled) before re-launching 1.35.1. When enabling LED control, this also fails with an import error (see attached log file). I tried starting entirely from scratch, re-imaging my SD card and installing again with the all in one script. However, the result and behavior was the same. Am I doing something wrong, or is there a different install method I should try?

[production_logs.log](https://github.com/user-attachments/files/16753370/production_logs.log)

**Comment by AndreWohnsland** [comment, marked as answer]
Hey, thanks for reaching out.
The update issue should be fixed, the next time you update after prompted. This was due to some issue within the project's metadata. You were up to date, but the code thought it was not, due to a mismatch.

For the LED part: How familiar are you with programming / Python or Linux in general?

In short, there is the `rpi_ws281x` library missing. There is also some doc for the LED at: https://cocktailberry.readthedocs.io/troubleshooting/#get-the-led-working.

Setting up my LEDs in the past, I had some issues, because the library needs access to places, usually not accessible without root rights (at least if nothing was altered). A quick fix at the time was to run the program as root, I also installed the dependencies into to python used by root **This is not recommended and more of a workaround!**

With the latest Raspberry Pi OS (or latest Debian in general) the recommendation is to use a venv, and that's what CocktailBerry also does. For starters, you can activate the venv and install the package there:
```bash
source ~/.env-cocktailberry/bin/activate
pip install rpi_ws281x
``` 
This fixes the problem, that the library is not is installed, but not the root privilege problem. Automating this is somewhere back in my todo list, but might take some intensive testing. So if you have some experience, you should probably be able to set the system privileges up to use with https://github.com/jgarff/rpi_ws281x, otherwise, I would recommend waiting a little if I can tackle this issue soon.

Update: I quick look at similar issues point out that PWM/PCM pins will probably not work without root, SPI (Phsical pin 10/ GPIO 10 for program) might just work out of the box.

**Comment by AtrociKitty** [comment]
I appreciate the quick reply. The update issue does look resolved, thanks for fixing that.

As for the LEDs, I initially suspected both this and the update issue were permissions-related, so I'm glad I was on the right path. With your direction, I now have LED control working using a WS2811 strip on GPIO 10. Aside from root access and the rpi_ws281x package, this also required that I enable SPI in the config file and remove any SPI pins from pump control (they may have worked, but there was a warning). I was also seeing a warning about wrong permissions on /run/user/1000/ that I'm not sure was related, but I addressed this as well.

I'm sure you have plenty to work on, but one other quick suggestion would be to consider making MAKER_TUBE_VOLUME a per pump setting, rather than global. I'm using two different pump models with two different tubing diameters in my build, so the difference in tube volume between pumps is fairly significant.

**Reply by AndreWohnsland** [reply]
Also, thanks for this additional information on how you resolved the LED issue. Might help my future me or others if there is a similar problem.

For the setting: I have somewhere in the back of my mind the thought of generally refactoring the config and how the auto-generation of the GUI works, so if I tackle this issue I am sure I can sneak such change also into the configuration.

**Reply by AndreWohnsland** [reply]
As an update: With https://github.com/AndreWohnsland/CocktailBerry/pull/184, which will be included in v1.36.0, there will be some improvements in the configuration GUI, including the split of `MAKER_TUBE_VOLUME` to be adjustable for each pump.
